java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java

```diff
@@ -0,0 +1,125 @@
+package net.osmand.server.api.services;
+
+import net.osmand.server.api.repo.DeviceSubscriptionsRepository;
+import net.osmand.server.api.repo.PremiumUsersRepository;
+import net.osmand.server.api.repo.PromoRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+@Service
+public class PromoService {
+    
+    @Autowired
+    PromoRepository promoRepository;
+    
+    @Autowired
+    private EmailSenderService emailSender;
+    
+    @Autowired
+    private PremiumUsersRepository usersRepository;
+    
+    @Autowired
+    protected UserSubscriptionService userSubService;
+    
+    @Autowired
+    private DeviceSubscriptionsRepository subscriptionsRepository;
+    
+    public void register(PromoRepository.Promo promo) {
+        if (promoRepository.existsById(promo.name)) {
+            throw new IllegalStateException("Promo already exists");
+        } else {
+            promoRepository.save(promo);
+        }
+    }
+    
+    public ResponseEntity<String> addUser(String name, String email) {
+        PromoRepository.Promo promo = promoRepository.findByName(name);
+        if (promo.used < promo.numberLimit) {
+            String key = "promo_" + promo.name;
+            Date expireTime = getExpirationDate(promo);
+            PromoResponse resp = createPromoSubscription(email, key, expireTime);
+            PremiumUsersRepository.PremiumUser existingUser = usersRepository.findByEmail(email);
+            if (existingUser != null && !resp.error) {
+                promo.used++;
```

Comment:
What if users will be registered concurrently ? should be the comment or the proper fix that field used could be bigger than limit?